event-stream-service/charts/event-stream-service/values.yaml

```
@@ -0,0 +1,204 @@
+          - user: "chefs"
+            password: << $CHEFS_PWD >>
+            permissions:
+              publish:
+                [
+                  "$JS.API.INFO",
+                  "$JS.API.STREAM.CREATE.CHEFS",
+                  "$JS.API.STREAM.UPDATE.CHEFS",
+                  "$JS.API.STREAM.DELETE.CHEFS",
+                  "$JS.API.STREAM.INFO.CHEFS",
+                  "$JS.API.STREAM.PURGE.CHEFS",
+                  "$JS.API.STREAM.LIST",
+                  "$JS.API.STREAM.NAMES",
+                  "$JS.API.STREAM.MSG.DELETE.CHEFS",
+                  "$JS.API.STREAM.MSG.GET.CHEFS",
+                  "$JS.API.STREAM.SNAPSHOT.CHEFS",
+                  "$JS.API.STREAM.RESTORE.CHEFS",
+                  "$JS.API.CONSUMER.CREATE.CHEFS",
+                  "$JS.API.CONSUMER.CREATE.CHEFS.>",
+                  "$JS.API.CONSUMER.DURABLE.CREATE.CHEFS.>",
+                  "$JS.API.CONSUMER.DELETE.CHEFS.>",
+                  "$JS.API.CONSUMER.INFO.CHEFS.>",
+                  "$JS.API.CONSUMER.LIST.CHEFS",
+                  "$JS.API.CONSUMER.NAMES.CHEFS",
+                  "$JS.API.CONSUMER.MSG.NEXT.CHEFS.>",
+                  "$JS.API.CONSUMER.MSG.NEXT.CHEFS.NEW",
+                  "$JS.API.STREAM.MSG.GET.CHEFS",
+                  "$JS.ACK.CHEFS.>",
+                  "PUBLIC.forms.>",
+                  "PRIVATE.forms.>",
+                ]
+              subscribe: ["_INBOX.>"]
+      no_auth_user: anonymous
+      01$include: params.conf
+  container:
+    env:
+      ADMIN_PWD:
+        valueFrom:
+          secretKeyRef:
+            name: ess-nats-auth
+            key: admin_pwd
+      CHEFS_PWD:
+        valueFrom:
+          secretKeyRef:
+            name: ess-nats-auth
+            key: chefs_pwd
+      ANONYMOUS_PWD:
+        valueFrom:
+          secretKeyRef:
+            name: ess-nats-auth
+            key: anonymous_pwd
+    merge:
+      resources:
+        limits:
+          cpu: 250m
+          memory: 128Mi
+        requests:
+          cpu: 50m
+          memory: 64Mi
+  service:
+    ports:
+      nats:
+        enabled: true
+      leafnodes:
+        enabled: false
+      websocket:
+        enabled: true
+      mqtt:
+        enabled: false
+      cluster:
+        enabled: true
+      gateway:
+        enabled: false
+      monitor:
+        enabled: false
+      profiling:
+        enabled: false
+  podTemplate:
+    topologySpreadConstraints:
+      kubernetes.io/hostname:
+        maxSkew: 1
+        whenUnsatisfiable: DoNotSchedule
+  configMap:
+    merge:
+      data:
+        params.conf: |
+          max_payload: 2MB
+  natsBox:
+    container:
+      image:
+        tag: nonroot
+      merge:
+        resources:
+          limits:
+            cpu: 250m
+            memory: 128Mi
+          requests:
+            cpu: 50m
+            memory: 64Mi
+
+nginx:
+  fullnameOverride: ess-nginx
+  readinessProbe:
+    enabled: true
+    path: /health
+  resourcesPreset: "nano"
```

For Nginx to be high availability during cluster maintenance, there needs to be a minimum of three pods.
```suggestion
    path: /health
  replicaCount: 3
  resourcesPreset: "nano"
```

---

placed in values-dev.yml

---

In dev I created a consumer and when I connect to the route it is correctly load balancing between the different nginx pods. However, the nginx pods all seem to be connecting to the same ess-nats-0. If I delete ess-nats-0, the consumer loses its connection and stops running. Restarting the consumer fails with a timeout until ess-nats-0 is again ready.

---

Similarly, trying to submit a form while ess-nats-0 is restarting causes an error in the CHEFS pod:
```
NatsError: 503
    at NatsError.errorForCode (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/core.js:143:16)
    at isRequestError (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/msg.js:29:33)
    at SubscriptionImpl.callback (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/muxsubscription.js:82:56)
    at ProtocolHandler.processMsg (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/protocol.js:615:17)
    at ProtocolHandler.push (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/protocol.js:722:22)
    at Parser.parse (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/parser.js:212:41)
    at ProtocolHandler.<anonymous> (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/protocol.js:426:45)
    at Generator.next (<anonymous>)
    at fulfilled (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/protocol.js:5:58)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Error
    at new BaseRequest (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/request.js:28:24)
    at new RequestOne (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/request.js:100:9)
    at NatsConnectionImpl.request (/opt/app-root/src/app/node_modules/nats.ws/lib/nats-base-client/nats.js:367:23)
    at JetStreamClientImpl.<anonymous> (/opt/app-root/src/app/node_modules/nats.ws/lib/jetstream/jsclient.js:138:39)
    at Generator.next (<anonymous>)
    at /opt/app-root/src/app/node_modules/nats.ws/lib/jetstream/jsclient.js:22:71
    at new Promise (<anonymous>)
    at __awaiter (/opt/app-root/src/app/node_modules/nats.ws/lib/jetstream/jsclient.js:18:12)
    at JetStreamClientImpl.publish (/opt/app-root/src/app/node_modules/nats.ws/lib/jetstream/jsclient.js:103:16)
    at EventStreamService._onSubmitEventStream (/opt/app-root/src/app/src/components/eventStreamService.js:322:39)
```

---

Thanks for this @WalterMoar. Yes, I am trying to figure out the "stickiness". Using `nats bench` was revealing the same issue (always ended up using the same pod, so can only store as many bytes as that pod's PVC can hold). And when that storage max was hit the pod would restart and the connection wouldn't move across to another pod... lot to investigate.

---

nginx is the proxy that makes sure the websockets can connect properly. as the client is negotiating the connection it expects certain headers which are set by nginx.

```
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "Upgrade";
```

---

OK, don't understand that, but FYI there are errors in the NATS pods like "websocket handshake error: invalid value for header 'Upgrade'"

---

i was in test, but was this around the time you were taking pods down? are there hundreds of these messages, is it consistently happening?